controller/src/main/java/io/pravega/controller/server/bucket/UniformBucketDistributor.java

```diff
@@ -0,0 +1,143 @@
+ * This is based only on the number of buckets running on a given controller and does not consider any other metric.
+ *
+ * The algorithm works as follows:
+ * - Sort the controllers based on the number of buckets they own.
+ * - Repeat the following until the controllers are balanced.
+ *     -- Remove a bucket from the controller containing max number of buckets.
+ *     -- Add this to the controller containing min number of buckets.
+ * - The balanced condition happens when the difference between min and max containers is at most 1.
+ */
+@Slf4j
+public class UniformBucketDistributor implements BucketDistributor {
+    @Override
+    public Map<String, Set<Integer>> distribute(Map<String, Set<Integer>> previousBucketControllerMapping,
+                                                Set<String> currentControllers, int bucketCount) {
+        Preconditions.checkNotNull(previousBucketControllerMapping, "previousBucketControllerMapping");
+        Preconditions.checkNotNull(currentControllers, "currentControllers");
+        if (currentControllers.isEmpty()) {
+            log.info("No controller found during distribution, creating empty map");
+            return new HashMap<>();
+        }
+        if (previousBucketControllerMapping.keySet().equals(currentControllers)) {
+            //Assuming the input map is always uniform distributed , since this distributor only depends on the controllers list.
+            log.debug("No change in controller list, using existing map");
+            return new HashMap<>(previousBucketControllerMapping);
+        }
+
+        if (previousBucketControllerMapping.size() == 0 || currentControllers.size() == 1) {
+            log.info("Creating new balanced map");
+            return initializeMap(currentControllers, bucketCount);
+        }
+
+        //Detect controllers not present in the existing Map.
+        Set<String> controllersAdded = currentControllers.stream().
+                                           filter(h -> !previousBucketControllerMapping.containsKey(h)).collect(Collectors.toSet());
+        if (!controllersAdded.isEmpty()) {
+            log.debug("New controllers added: {}", controllersAdded);
+        }
+        //Detect all buckets whose controllers has been removed. These need to be assigned to new controllers.
+        Set<Integer> orphanedBuckets = previousBucketControllerMapping.entrySet().stream().
+                                                             filter(h -> !currentControllers.contains(h.getKey()))
+                                                                   .flatMap(p -> p.getValue().stream())
+                                                                   .collect(Collectors.toSet());
+
+        //Using a TreeSet sorted by number of containers to optimize add/remove operations.
+        TreeSet<Map.Entry<String, Set<Integer>>> mapElements =
```

Comment:
Is it true that this algorithm will keep bucket movements to the minimum, right? I mean, if there is some assignment of buckets to controllers, the algorithm will only move some containers from the most loaded controllers to the new controller, but all other assignments will remain the same, right?

Reply:
Yes. This algorithm ensure us number of buckets will remains uniform among available controllers. Also it ensure us there will be minimum shuffling of buckets while redistribution.